From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: huangqiqi <huangqiqi@loongson.cn>
Date: Mon, 21 Oct 2024 11:12:22 +0800
Subject: [PATCH 33/91] LOONGARCH64: Optimize LOONGF_I to LOONGF_I* to
 eliminate the use of i&0xf...

LOONGF_I => LOONGF_I5, LOONGF_I6, LOONGF_I12, LOONGF_I16, LOONGF_I20,
	    LOONGF_I21, LOONGF_I26
---
 src/lj_asm_loongarch64.h    | 52 ++++++++++++++++++-------------------
 src/lj_ccallback.c          | 22 ++++++++--------
 src/lj_emit_loongarch64.h   | 42 +++++++++++++++---------------
 src/lj_target_loongarch64.h | 12 ++++++---
 4 files changed, 67 insertions(+), 61 deletions(-)

diff --git a/src/lj_asm_loongarch64.h b/src/lj_asm_loongarch64.h
index 094d8e1b..9d9d6839 100644
--- a/src/lj_asm_loongarch64.h
+++ b/src/lj_asm_loongarch64.h
@@ -53,23 +53,23 @@ static void asm_exitstub_setup(ASMState *as)
   mxp--;
   ptrdiff_t delta = ((uintptr_t)(void *)lj_vm_exit_handler - (uintptr_t)mxp)>>2;
   if (LOONGF_S_OK(delta, 26)) {
-    *mxp = LOONGI_B | LOONGF_I(delta & 0xffffu) | ((delta & 0x3ff0000) >> 16);
+    *mxp = LOONGI_B | LOONGF_I26(delta);
   } else {
-    *mxp = LOONGI_JIRL | RID_R0 | LOONGF_J(RID_TMP) | 0<<10;
+    *mxp = LOONGI_JIRL | LOONGF_D(RID_R0) | LOONGF_J(RID_TMP) | LOONGF_I16(0);
     uintptr_t target = (uintptr_t)(void *)lj_vm_exit_handler;
-    *--mxp = LOONGI_LU52I_D | RID_TMP | LOONGF_J(RID_TMP) | LOONGF_I((target>>52)&0xfff);
-    *--mxp = LOONGI_LU32I_D | RID_TMP | LOONGF_I20((target>>32)&0xfffff);
-    *--mxp = LOONGI_ORI | RID_TMP | LOONGF_J(RID_TMP) | LOONGF_I(target&0xfff);
-    *--mxp = LOONGI_LU12I_W | RID_TMP | LOONGF_I20((target&0xfffff000)>>12);
+    *--mxp = LOONGI_LU52I_D | LOONGF_D(RID_TMP) | LOONGF_J(RID_TMP) | LOONGF_I12(target>>52);
+    *--mxp = LOONGI_LU32I_D | LOONGF_D(RID_TMP) | LOONGF_I20(target>>32);
+    *--mxp = LOONGI_ORI | LOONGF_D(RID_TMP) | LOONGF_J(RID_TMP) | LOONGF_I12(target);
+    *--mxp = LOONGI_LU12I_W | LOONGF_D(RID_TMP) | LOONGF_I20(target>>12);
   }
-  *--mxp = LOONGI_ST_W | LOONGF_D(RID_TMP) | LOONGF_J(RID_SP) | LOONGF_I(4);
+  *--mxp = LOONGI_ST_W | LOONGF_D(RID_TMP) | LOONGF_J(RID_SP) | LOONGF_I12(4);
   if (checki12(as->T->traceno)) {
-    *--mxp = LOONGI_ADDI_D | LOONGF_D(RID_TMP) | RID_ZERO | LOONGF_I(as->T->traceno&0xfff);
+    *--mxp = LOONGI_ADDI_D | LOONGF_D(RID_TMP) | RID_ZERO | LOONGF_I12(as->T->traceno);
   } else {
-    *--mxp = LOONGI_ORI | LOONGF_D(RID_TMP) | LOONGF_J(RID_TMP) | LOONGF_I(as->T->traceno&0xfff);
-    *--mxp = LOONGI_LU12I_W | LOONGF_D(RID_TMP) | LOONGF_I20((as->T->traceno>>12)&0xfffff);
+    *--mxp = LOONGI_ORI | LOONGF_D(RID_TMP) | LOONGF_J(RID_TMP) | LOONGF_I12(as->T->traceno);
+    *--mxp = LOONGI_LU12I_W | LOONGF_D(RID_TMP) | LOONGF_I20(as->T->traceno>>12);
   }
-  *--mxp = LOONGI_ST_W | LOONGF_D(RID_TMP) | LOONGF_J(RID_SP) | LOONGF_I(0);
+  *--mxp = LOONGI_ST_W | LOONGF_D(RID_TMP) | LOONGF_J(RID_SP) | LOONGF_I12(0);
   as->mctop = mxp;
 }
 
@@ -331,7 +331,7 @@ static void asm_callx(ASMState *as, IRIns *ir)
     ci.func = (ASMFunction)(void *)get_kval(as, func);
   } else {  /* Need specific register for indirect calls. */
     Reg freg = ra_alloc1(as, func, RSET_RANGE(RID_R12, RID_MAX_GPR)&(~RSET_FIXED));
-    *--as->mcp = LOONGI_JIRL | LOONGF_D(RID_RA) | LOONGF_J(freg);
+    emit_dj(as, LOONGI_JIRL, RID_RA, freg);
     ci.func = (ASMFunction)(void *)0;
   }
   asm_gencall(as, &ci, args);
@@ -795,7 +795,7 @@ static void asm_href(ASMState *as, IRIns *ir, IROp merge)
     emit_loadi(as, RID_TMP, as->snapno);
   }
   emit_djs12(as, LOONGI_LD_D, tmp1, dest, offsetof(Node, key.u64));
-  *l_loop = LOONGI_BNE | LOONGF_J(tmp1) | LOONGF_D(RID_ZERO) | LOONGF_I(((as->mcp-l_loop) & 0xffffu));
+  *l_loop = LOONGI_BNE | LOONGF_J(tmp1) | LOONGF_D(RID_ZERO) | LOONGF_I16(as->mcp-l_loop);
   if (!isk && irt_isaddr(kt)) {
     type = ra_allock(as, (int64_t)irt_toitype(kt) << 47, allow);
     emit_djk(as, LOONGI_ADD_D, tmp2, key, type);
@@ -1790,16 +1790,16 @@ static void asm_loop_fixup(ASMState *as)
     uint32_t mask = (p[-2] & 0xfc000000) == 0x48000000 ? 0x1fffffu : 0xffffu;
     ptrdiff_t delta = target - (p - 2);
     if (mask == 0x1fffffu) {  /* BCEQZ  BCNEZ*/
-      p[-2] = p[-2] | LOONGF_I((uint32_t)delta & 0xffffu) | (((uint32_t)delta & 0x1f0000u) >> 16);
+      p[-2] = p[-2] | LOONGF_I21(delta);
     } else {  /* BEQ BNE BLE BGE BLTU BGEU*/
-      p[-2] |= LOONGF_I(delta & 0xffffu);
+      p[-2] |= LOONGF_I16(delta);
     }
     if (p[-1] == 0)
       p[-1] = LOONGI_NOP;
   } else {
     /* b */
     ptrdiff_t delta = target - (p - 1);
-    p[-1] = LOONGI_B | LOONGF_I(delta & 0xffffu) | ((delta & 0x3ff0000) >> 16);
+    p[-1] = LOONGI_B | LOONGF_I26(delta);
   }
 }
 
@@ -1857,11 +1857,11 @@ static void asm_tail_fixup(ASMState *as, TraceNo lnk)
   if (spadj == 0) {
     p[-1] = LOONGI_NOP;
   } else {
-    p[-1] = LOONGI_ADDI_D|LOONGF_D(RID_SP)|LOONGF_J(RID_SP)|LOONGF_I(spadj);
+    p[-1] = LOONGI_ADDI_D|LOONGF_D(RID_SP)|LOONGF_J(RID_SP)|LOONGF_I12(spadj);
   }
 
   MCode *tmp = p;
-  *p = LOONGI_B | LOONGF_I((uintptr_t)(target-tmp)&0xffffu) | (((uintptr_t)(target-tmp)&0x3ff0000u) >> 16);
+  *p = LOONGI_B | LOONGF_I26((uintptr_t)(target-tmp));
 }
 
 /* Prepare tail of code. */
@@ -1932,7 +1932,7 @@ void lj_asm_patchexit(jit_State *J, GCtrace *T, ExitNo exitno, MCode *target)
   MCode *cstart = NULL, *cstop = NULL;
   MCode *mcarea = lj_mcode_patch(J, p, 0);
 
-  MCode exitload = LOONGI_ADDI_D | LOONGF_D(RID_TMP) | LOONGF_J(RID_ZERO) | LOONGF_I(exitno&0xfff);
+  MCode exitload = LOONGI_ADDI_D | LOONGF_D(RID_TMP) | LOONGF_J(RID_ZERO) | LOONGF_I12(exitno);
 
   for (; p < pe; p++) {
     if (*p == exitload) {
@@ -1949,7 +1949,7 @@ void lj_asm_patchexit(jit_State *J, GCtrace *T, ExitNo exitno, MCode *target)
 	  /* nothing */
 	  continue;
         } else if (LOONGF_S_OK(delta, 16)) {
-          p[1] = (ins & 0xfc0003ffu) | LOONGF_I(delta & 0xffff);
+          p[1] = (ins & 0xfc0003ffu) | LOONGF_I16(delta);
           *p = LOONGI_NOP;
         } else if (LOONGF_S_OK(delta, 21)) {
 	  Reg rj = (ins>>5) & 0x1f;
@@ -1957,19 +1957,19 @@ void lj_asm_patchexit(jit_State *J, GCtrace *T, ExitNo exitno, MCode *target)
 	  switch (ins & 0xfc000000u) {
           case LOONGI_BEQ:
             *p = LOONGI_SUB_D | LOONGF_D(RID_TMP) | LOONGF_J(rj) | LOONGF_K(rd);
-            p[1] = LOONGI_BEQZ | LOONGF_J(RID_TMP) | LOONGF_I(delta & 0xffff) | ((delta & 0x1f0000) >> 16);
+            p[1] = LOONGI_BEQZ | LOONGF_J(RID_TMP) | LOONGF_I21(delta);
 	    break;
 	  case LOONGI_BNE:
             *p = LOONGI_SUB_D | LOONGF_D(RID_TMP) | LOONGF_J(rj) | LOONGF_K(rd);
-            p[1] = LOONGI_BNEZ | LOONGF_J(RID_TMP) | LOONGF_I(delta & 0xffff) | ((delta & 0x1f0000) >> 16);
+            p[1] = LOONGI_BNEZ | LOONGF_J(RID_TMP) | LOONGF_I21(delta);
 	    break;
 	  case LOONGI_BLT:
             *p = LOONGI_SLT | LOONGF_D(RID_TMP) | LOONGF_J(rj) | LOONGF_K(rd);
-            p[1] = LOONGI_BNEZ | LOONGF_J(RID_TMP) | LOONGF_I(delta & 0xffff) | ((delta & 0x1f0000) >> 16);
+            p[1] = LOONGI_BNEZ | LOONGF_J(RID_TMP) | LOONGF_I21(delta);
 	    break;
 	  case LOONGI_BGE:
             *p = LOONGI_SLT | LOONGF_D(RID_TMP) | LOONGF_J(rj) | LOONGF_K(rd);
-            p[1] = LOONGI_BEQZ | LOONGF_J(RID_TMP) | LOONGF_I(delta & 0xffff) | ((delta & 0x1f0000) >> 16);
+            p[1] = LOONGI_BEQZ | LOONGF_J(RID_TMP) | LOONGF_I21(delta);
 	    break;
 	  }
 	} else {
@@ -1981,7 +1981,7 @@ void lj_asm_patchexit(jit_State *J, GCtrace *T, ExitNo exitno, MCode *target)
                  ((ins & 0xfc000000u) == LOONGI_BCEQZ ||
                   (ins & 0xfc000100u) == LOONGI_BCNEZ)) {
         if (LOONGF_S_OK(delta, 21)) {
-          p[1] = (ins & 0xfc0003e0u) | LOONGF_I(delta & 0xffff) | ((delta & 0x1f0000) >> 16);
+          p[1] = (ins & 0xfc0003e0u) | LOONGF_I21(delta);
           *p = LOONGI_NOP;
 	  cstop = p + 2;
           if (!cstart) cstart = p;
@@ -1992,7 +1992,7 @@ void lj_asm_patchexit(jit_State *J, GCtrace *T, ExitNo exitno, MCode *target)
          if (p[2] == LOONGI_NOP) {
             ptrdiff_t delta = target - &p[2];
             lj_assertJ(LOONGF_S_OK(delta, 26), "branch target out of range");
-            p[2] = LOONGI_B | LOONGF_I(delta & 0xffff) | ((delta & 0x3ff0000) >> 16);
+            p[2] = LOONGI_B | LOONGF_I26(delta);
             *p = LOONGI_NOP;
 	    cstop = p + 3;
             if (!cstart) cstart = p + 2;
diff --git a/src/lj_ccallback.c b/src/lj_ccallback.c
index ffbfaf81..919be275 100644
--- a/src/lj_ccallback.c
+++ b/src/lj_ccallback.c
@@ -308,22 +308,22 @@ static void *callback_mcode_init(global_State *g, uint32_t *page)
   uintptr_t target = (uintptr_t)(void *)lj_vm_ffi_callback;
   uintptr_t ug = (uintptr_t)(void *)g;
   MSize slot;
-  *p++ = LOONGI_LU12I_W | LOONGF_D(RID_R18) | LOONGF_I20((target >> 12) & 0xfffff);
-  *p++ = LOONGI_LU12I_W | LOONGF_D(RID_R17) | LOONGF_I20((ug >> 12) & 0xfffff);
-  *p++ = LOONGI_ORI  | LOONGF_D(RID_R18) | LOONGF_J(RID_R18) | LOONGF_I(target & 0xfff);
-  *p++ = LOONGI_ORI  | LOONGF_D(RID_R17) | LOONGF_J(RID_R17) | LOONGF_I(ug & 0xfff);
-  *p++ = LOONGI_LU32I_D | LOONGF_D(RID_R18) | LOONGF_I20((target >> 32) & 0xfffff);
-  *p++ = LOONGI_LU32I_D | LOONGF_D(RID_R17) | LOONGF_I20((ug >> 32) & 0xfffff);
-  *p++ = LOONGI_LU52I_D | LOONGF_D(RID_R18) | LOONGF_J(RID_R18) | LOONGF_I((target >> 52) & 0xfff);
-  *p++ = LOONGI_LU52I_D | LOONGF_D(RID_R17) | LOONGF_J(RID_R17) | LOONGF_I((ug >> 52) & 0xfff);
+  *p++ = LOONGI_LU12I_W | LOONGF_D(RID_R18) | LOONGF_I20(target >> 12);
+  *p++ = LOONGI_LU12I_W | LOONGF_D(RID_R17) | LOONGF_I20(ug >> 12);
+  *p++ = LOONGI_ORI  | LOONGF_D(RID_R18) | LOONGF_J(RID_R18) | LOONGF_I12(target);
+  *p++ = LOONGI_ORI  | LOONGF_D(RID_R17) | LOONGF_J(RID_R17) | LOONGF_I12(ug);
+  *p++ = LOONGI_LU32I_D | LOONGF_D(RID_R18) | LOONGF_I20(target >> 32);
+  *p++ = LOONGI_LU32I_D | LOONGF_D(RID_R17) | LOONGF_I20(ug >> 32);
+  *p++ = LOONGI_LU52I_D | LOONGF_D(RID_R18) | LOONGF_J(RID_R18) | LOONGF_I12(target >> 52);
+  *p++ = LOONGI_LU52I_D | LOONGF_D(RID_R17) | LOONGF_J(RID_R17) | LOONGF_I12(ug >> 52);
   *p++ = LOONGI_NOP;
   *p++ = LOONGI_NOP;
   *p++ = LOONGI_NOP;
   *p++ = LOONGI_NOP;
-  *p++ = LOONGI_JIRL | LOONGF_D(RID_R0) | LOONGF_J(RID_R18) | LOONGF_I(0);
+  *p++ = LOONGI_JIRL | LOONGF_D(RID_R0) | LOONGF_J(RID_R18) | LOONGF_I16(0);
   for (slot = 0; slot < CALLBACK_MAX_SLOT; slot++) {
-    *p++ = LOONGI_ORI  | LOONGF_D(RID_R19) | LOONGF_J(RID_R0) | LOONGF_I(slot & 0xfff);
-    *p = LOONGI_B | LOONGF_I((page-p) & 0xffff) | (((page-p) >> 16) & 0x3ff);
+    *p++ = LOONGI_ORI  | LOONGF_D(RID_R19) | LOONGF_J(RID_R0) | LOONGF_I12(slot);
+    *p = LOONGI_B | LOONGF_I26(page-p);
     p++;
   }
   return p;
diff --git a/src/lj_emit_loongarch64.h b/src/lj_emit_loongarch64.h
index 05499409..c6b2004b 100644
--- a/src/lj_emit_loongarch64.h
+++ b/src/lj_emit_loongarch64.h
@@ -26,39 +26,39 @@ static intptr_t get_k64val(ASMState *as, IRRef ref)
 
 static void emit_djk(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, Reg rk)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_K(rk & 0x1f);
+  *--as->mcp = loongi | LOONGF_D(rd&0x1f) | LOONGF_J(rj&0x1f) | LOONGF_K(rk&0x1f);
 }
 
 #define emit_dj(as, loongi, rd, rj)         emit_djk(as, loongi, rd, rj, 0)
 
 static void emit_dju5(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, uint32_t u)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(u&0x1f);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I5(u);
 }
 
 static void emit_dju6(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, uint32_t u)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(u&0x3f);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I6(u);
 }
 
 static void emit_djs12(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, int32_t i)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(i&0xfff);
+  *--as->mcp = loongi | LOONGF_D(rd&0x1f) | LOONGF_J(rj) | LOONGF_I12(i);
 }
 
 static void emit_dju12(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, uint32_t u)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(u&0xfff);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I12(u);
 }
 
 static void emit_djs16(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, int32_t i)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(i&0xffff);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I16(i);
 }
 
 static void emit_ds20(ASMState *as, LOONGIns loongi, Reg rd, int32_t i)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_I20(i & 0xfffff);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_I20(i);
 }
 
 #define checki12(x)	LOONGF_S_OK(x, 12)
@@ -70,17 +70,17 @@ static Reg ra_scratch(ASMState *as, RegSet allow);
 
 static void emit_djml(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, uint32_t m, uint32_t l)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_I(l & 0x3f) | LOONGF_M(m & 0x3f);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_L(l) | LOONGF_M(m);
 }
 
 static void emit_djka(ASMState *as, LOONGIns loongi, Reg rd, Reg rj, Reg rk, Reg ra)
 {
-  *--as->mcp = loongi | LOONGF_D(rd & 0x1f) | LOONGF_J(rj & 0x1f) | LOONGF_K(rk & 0x1f) | LOONGF_A(ra & 0x1f);
+  *--as->mcp = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_K(rk) | LOONGF_A(ra);
 }
 
 static void emit_b_bl(ASMState *as, LOONGIns loongi, uint32_t i)
 {
-  *--as->mcp = loongi | LOONGF_I(i & 0xffff) | ((i >> 16) & 0x3ff);
+  *--as->mcp = loongi | LOONGF_I26(i);
 }
 
 
@@ -101,7 +101,7 @@ static void emit_loads32(ASMState *as, Reg r, int32_t i)
 static void emit_loadi(ASMState *as, Reg r, int32_t i)
 {
   if (checki12(i)) {
-    *--as->mcp = LOONGI_ADDI_D | LOONGF_D(r&0x1f) | LOONGF_J(RID_ZERO) | LOONGF_I(i&0xfff);
+    *--as->mcp = LOONGI_ADDI_D | LOONGF_D(r) | LOONGF_J(RID_ZERO) | LOONGF_I12(i);
   } else {
     emit_loads32(as, r, i);
   }
@@ -113,10 +113,10 @@ static void emit_loadu64(ASMState *as, Reg r, uint64_t u64)
   if (checki32((int64_t)u64)) {
     emit_loadi(as, r, (int32_t)u64);
   } else {
-    *--as->mcp = LOONGI_LU52I_D | LOONGF_D(r) | LOONGF_J(r) | LOONGF_I((u64>>52)&0xfff);
-    *--as->mcp = LOONGI_LU32I_D | LOONGF_D(r) | LOONGF_I20((u64>>32)&0xfffff);
-    *--as->mcp = LOONGI_ORI | LOONGF_D(r) | LOONGF_J(r) | LOONGF_I(u64&0xfff);
-    *--as->mcp = LOONGI_LU12I_W | LOONGF_D(r) | LOONGF_I20((u64>>12)&0xfffff);
+    *--as->mcp = LOONGI_LU52I_D | LOONGF_D(r) | LOONGF_J(r) | LOONGF_I12(u64>>52);
+    *--as->mcp = LOONGI_LU32I_D | LOONGF_D(r) | LOONGF_I20(u64>>32);
+    *--as->mcp = LOONGI_ORI | LOONGF_D(r) | LOONGF_J(r) | LOONGF_I12(u64);
+    *--as->mcp = LOONGI_LU12I_W | LOONGF_D(r) | LOONGF_I20(u64>>12);
   }
 }
 
@@ -195,7 +195,7 @@ static void emit_branch(ASMState *as, LOONGIns loongi, Reg rj, Reg rd, MCode *ta
   ptrdiff_t delta = target - (p - 1);
   lj_assertA(((delta + 0x8000) >> 16) == 0, "branch target out of range");
   /* BEQ BNE BGE BLZ */
-  *--p = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I(((uint32_t)delta & 0xffffu));
+  *--p = loongi | LOONGF_D(rd) | LOONGF_J(rj) | LOONGF_I16(delta);
   as->mcp = p;
 }
 
@@ -204,8 +204,8 @@ static void emit_branch21(ASMState *as, LOONGIns loongi, Reg rj, MCode *target)
   MCode *p = as->mcp;
   ptrdiff_t delta = target - (p - 1);
   lj_assertA(((delta + 0x100000) >> 21) == 0, "branch target out of range");
-  *--p = loongi | LOONGF_J(rj) | LOONGF_I(((uint32_t)delta & 0xffffu))
-         | (((uint32_t)delta & 0x1f0000u)>>16);  /* BEQZ BNEZ BCEQZ BCNEZ */
+  /* BEQZ BNEZ BCEQZ BCNEZ */
+  *--p = loongi | LOONGF_J(rj) | LOONGF_I21(delta);
   as->mcp = p;
 }
 
@@ -213,7 +213,7 @@ static void emit_jmp(ASMState *as, MCode *target)
 {
   MCode *p = as->mcp;
   ptrdiff_t delta = target - (p - 1);
-  emit_b_bl(as, LOONGI_B, (delta&0x3ffffff));  /* offs 26 */
+  emit_b_bl(as, LOONGI_B, delta);  /* offs 26 */
 }
 
 #define emit_move(as, dst, src) \
@@ -224,10 +224,10 @@ static void emit_call(ASMState *as, void *target)
   MCode *p = --as->mcp;
   ptrdiff_t delta = (char *)target - (char *)p;
   if (LOONGF_S_OK(delta>>2, 26)) {
-    *p = LOONGI_BL | LOONGF_I((delta>>2) & 0xffff) | (((delta>>2) >> 16) & 0x3ff);
+    *p = LOONGI_BL | LOONGF_I26(delta>>2);
   } else {  /* Target out of range: need indirect call. */
     Reg r = ra_allock(as, (intptr_t)target, RSET_RANGE(RID_R12, RID_R19+1));
-    *p = LOONGI_JIRL | LOONGF_D(RID_RA) | LOONGF_J(r) | LOONGF_I(0);
+    *p = LOONGI_JIRL | LOONGF_D(RID_RA) | LOONGF_J(r) | LOONGF_I16(0);
   }
 }
 
diff --git a/src/lj_target_loongarch64.h b/src/lj_target_loongarch64.h
index 044d3cb9..5247b8a6 100644
--- a/src/lj_target_loongarch64.h
+++ b/src/lj_target_loongarch64.h
@@ -126,9 +126,15 @@ static LJ_AINLINE uint32_t *exitstub_trace_addr_(uint32_t *p)
 #define LOONGF_J(r)	((r) << 5)
 #define LOONGF_K(r)	((r) << 10)
 #define LOONGF_A(r)	((r) << 15)
-#define LOONGF_I(n)	((n) << 10)
-#define LOONGF_I20(n)	((n) << 5)
-#define LOONGF_M(n)	((n) << 16)
+#define LOONGF_I5(n)	(((uint32_t)(n) & 0x1fu) << 10)
+#define LOONGF_I6(n)	(((uint32_t)(n) & 0x3fu) << 10)
+#define LOONGF_I12(n)	(((uint32_t)(n) & 0xfffu) << 10)
+#define LOONGF_I16(n)	(((uint32_t)(n) & 0xffffu) << 10)
+#define LOONGF_I20(n)	(((uint32_t)(n) & 0xfffffu) << 5)
+#define LOONGF_I21(n)	((LOONGF_I16(n)) | (((uint32_t)(n) >> 16) & 0x1fu))
+#define LOONGF_I26(n)	((LOONGF_I16(n)) | (((uint32_t)(n) >> 16) & 0x3ffu))
+#define LOONGF_M(n)	(((uint32_t)(n) & 0x3fu) << 16)
+#define LOONGF_L(n)	(((uint32_t)(n) & 0x3fu) << 10)
 
 /* Check for valid field range. */
 #define LOONGF_S_OK(x, b) ((((x) + (1 << (b-1))) >> (b)) == 0)
